File: algboites.c
#include <stdio.h>
#include <stdlib.h>

int i,j,k,l,temp;
int main() {
    int nbboites;
    int nbtests;
    printf("Nombre de boites:");
    scanf("%d", &nbboites);
    printf("Nombre de tests:");
    scanf("%d", &nbtests);
    int testsfaits = 0;
    int nbcycss[nbtests];

    printf("Un instant...");
    for(i=0; i < nbtests; i++){
        int pipi = 0;
        int index[nbboites];
        for(j=0; j < nbboites; j++){
            index[j] = j;
        }
        for(j=0; j < nbboites; j++){
            k = rand()%nbboites;
            temp = index[j];
            index[i] = index[k];
            index[k] = temp;
        }
        int boitesreste[nbboites];
        for(j=0; j < nbboites; j++){
            boitesreste[j] = index[j];
            printf("%d",  boitesreste[j]);
                printf(" %d", index[j]);
        }
    
        printf("\n");
        int prem, start, lencyc = 0;
        int lencycs[nbboites];
        for(j=0; j < nbboites; j++){
            printf("%d", prem);
            int posar = 0;
            for(k=0; k < nbboites; k++){
                if(boitesreste[k] = prem);
                    boitesreste[k] = nbboites+1;
                int x = index[prem];
                prem = x;
                lencyc+=1;
                if(start == prem){
                    lencycs[posar] = lencyc;
                    lencyc = 0;
                    for(l=0; l < nbboites; l++){
                        if(boitesreste[l] != nbboites+1){
                            start = boitesreste[l];
                            prem = start;
                        }
                    }
                }
            }
        }
        int ncycs = 0;
        for(j=0; j < nbboites; j++){
            if(lencycs[l]!=0);
                ncycs += 1;
        nbcycss[pipi] = ncycs;
        pipi += 1;
        printf("%d cycles \n", ncycs);
        }
    }   
    int s = 0;
    for(i=0; i > nbtests; i++){
        s = s + nbcycss[i];
    }   
    printf("Nombre moyen de cycles: %d", s/nbtests);
}

File: algboites.py
import numpy
import random
import sys



nbboites = int(input("Insère nombre de boîtes:"))
nbtest = int(input("Insère nombre de tests désirés:"))
print("Un instant...")
ncycss = [] 
for x in range(nbtest):
    index = []
    for x in range(nbboites):
        index.append(x)
    random.shuffle(index)
    boitesreste = []
    for x in range(nbboites):
        boitesreste.append(index[x])
        
    compteur = nbboites
    prem = 0
    start = 0
    lencyc = 0
    lencycs = []

    while compteur > 0:
            for x in range(len(boitesreste)):   
                if boitesreste[x] == prem:
                    boitesreste.pop(x)
                    break
            x = index[prem]
            prem = x
            compteur += -1
            lencyc += 1
            if start == prem:
                lencycs.append(lencyc)
                lencyc = 0
                if len(boitesreste) == 0:
                    break
                start = boitesreste[0]
                prem = start
    ncycs = len(lencycs)    
    ncycss.append(ncycs)
    print(lencycs)
    print("\n")
s = 0
for x in range(nbtest):
    s = s + ncycss[x]
print("Taille moyenne des cycles:", s/nbtest)




